Functional tests for floresta-wire/sync-node

lla-dane opened the pull request:
Simulated a peer to take in requests for the sync-node instance. Filled the chain-state with some block-headers prior block validation and then send the appropriate responses to the node.

lla-dane commented:
@Davidson-Souza, please suggest some more scenarios to test like you did in chain-selector, if needed.

Davidson-Souza commented:
> @Davidson-Souza, please suggest some more scenarios to test like you did for chain-selector, if needed.

The sync node is very simple, so I don't think there's much to test. But some things I can think of:

 - Test how it responds to a node sending a block without proof (this should ban the node)
 - How it responds to a block with invalid proof (this should ban the node, but should **not** invalidate the block)
 - Sending blocks out-of-order (it should handle them fine)
 - Sending duplicated blocks (this will increase the peer's ban score)

Davidson-Souza commented:
Btw. The `test_utils` mod is the same code from #180 right? If you prefer, you can make one PR creating a module `floresta-wire/src/p2p-wire/tests/util.rs` and then rebase both this and #180 on that.

lla-dane commented:
> Btw. The `test_utils` mod is the same code from #180 right? If you prefer, you can make one PR creating a module `floresta-wire/src/p2p-wire/tests/util.rs` and then rebase both this and #180 on that.

Sure, I will do that.

lla-dane commented:
> The sync node is very simple, so I don't think there's much to test. But some things I can think of:
> 
> * Test how it responds to a node sending a block without proof (this should ban the node)
> * How it responds to a block with invalid proof (this should ban the node, but should **not** invalidate the block)
> * Sending blocks out-of-order (it should handle them fine)

Added these tests, and fixed a bug along the way #203. After merge of #203, I will rebase this.
@Davidson-Souza please take a look, and suggest if I have to change something, then I will squash these commits.

lla-dane commented:
phew, I guess I am done with this PR, please take a look @Davidson-Souza !!

Davidson-Souza commented on crates/floresta-wire/src/p2p_wire/tests/sync_node.rs:
```diff
@@ -0,0 +1,170 @@
+        // HENCE THE CHAIN WILL HAVE A HEIGHT OF 6.
+
+        // THIS SIMULATION WILL TEST:
+        // 1) SENDING BLOCK WITH INVALID PROOF: 7TH BLOCK IS SET AS INVALID.
```
Is the proof invalid, or the actual block? If the proof is invalid, we don't invalidate a block because this is not PoW committed and could easily be forged by peers.

lla-dane replied:
My bad here, I tweaked with the transactions so currently it is giving `BadMerkelRoot`, in the `self.validate_block` function in `connect_block`. So its invalidating the block.

Should I let it be and change the comments accordingly or ... ??

lla-dane replied:
But regardless the `InvalidProof` message, comes from the `validate_block` fucntion in the `BlockchainInterface` trait here: 
```rust
fn validate_block(
        &self,
        block: &Block,
        proof: Proof,
        inputs: HashMap<OutPoint, TxOut>,
        del_hashes: Vec<sha256::Hash>,
        acc: Stump,
    ) -> Result<(), Self::Error> {
        // verify the proof
        let del_hashes = del_hashes
            .iter()
            .map(|hash| NodeHash::from(hash.as_byte_array()))
            .collect::<Vec<_>>();

        if !acc.verify(&proof, &del_hashes)? {
            return Err(BlockValidationErrors::InvalidProof.into());
        }

        let height = self
            .get_block_height(&block.block_hash())?
            .ok_or(BlockchainError::BlockNotPresent)?;

        self.validate_block(block, height, inputs)
    }
```
but we are executing the `validate_block` function in the `ChainState` trait in `connect_block` function. 

@Davidson-Souza

Davidson-Souza replied:
> Should I let it be and change the comments accordingly or ... ??

Yes, please

> but we are executing the validate_block function in the ChainState trait in connect_block function. So where will the InvalidProof error message come from if not from BlokchainInterface ??

Invalid proof comes from `connect_block`, but we can't invalidate the block because of an invalid proof. This is due to proofs not being committed in the block itself, so a dishonest peer can produce an invalid proof without invalidating the PoW and make us invalidate that block, throwing us out of the network. Maybe you can add a test with this case: Some peer sends a block with invalid proof, we ban this peer and ask another. The second peer gives us a valid proof and we accept that block.

lla-dane commented on crates/floresta-wire/src/p2p_wire/mod.rs:
```diff
@@ -87,3 +87,4 @@ pub mod running_node;
 pub mod socks;
 pub mod stream_reader;
 pub mod sync_node;
+pub mod tests;
```
Done !!
